packages/@react-facet/core/src/hooks/useFacetCallback.ts

```diff
@@ -1,47 +1,158 @@
+    Facet<V5>,
+    Facet<V6>,
+    Facet<V7>,
+    Facet<V8>,
+    Facet<V9>,
+    Facet<V10>,
+  ],
+): C
 
 /**
  * Creates a callback that depends on the value of a facet.
  * Very similar to `useCallback` from `React`
  *
- * @param callback function callback that receives the current facet value and the arguments passed to the callback
- * @param dependencies variable used by the map that are available in scope (similar as dependencies of useCallback)
- * @param facet facet that the callback listens to
+ * @param callback function callback that receives the current facet values and the arguments passed to the callback
+ * @param dependencies variable used by the callback that are available in scope (similar as dependencies of useCallback)
+ * @param facets facets that the callback listens to
  *
  * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
  * Having this as the second argument allows the linter to work.
  */
-export function useFacetCallback<V, T>(
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  callback: (value: V) => (...args: any[]) => T | NoValue,
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  dependencies: any[],
-  facet: Facet<V>,
-) {
-  const facetRef = useRef<Option<V>>(facet.get())
-
-  useFacetEffect(
-    (value) => {
-      facetRef.current = value
-    },
-    [],
-    facet,
-  )
+export function useFacetCallback<M, C extends (...args: unknown[]) => M | NoValue>(
+  callback: (...args: unknown[]) => C,
+  dependencies: unknown[],
+  facets: Facet<unknown>[],
+): C {
+  const facetsRef = useRef<Option<unknown>[]>(facets.map(() => NO_VALUE))
+
+  useEffect(() => {
+    const unsubscribe = facets.map((facet, index) =>
+      facet.observe((value) => {
+        facetsRef.current[index] = value
+      }),
+    )
+
+    return () => {
+      unsubscribe.forEach((unsubscribe) => unsubscribe())
+    }
+    // We care about each individual facet and if any is a different reference
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, facets)
+
+  // We care about each individual dependency and if any is a different reference
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const callbackMemoized = useCallback(callback, dependencies)
 
-  return useCallback(
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  return useCallback<C>(
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    (...args: any[]): T | NoValue => {
-      const value = facetRef.current
+    ((...args: any[]) => {
+      const values = facetsRef.current
 
-      if (value === NO_VALUE) return NO_VALUE
+      for (const value of values) {
+        if (value === NO_VALUE) return NO_VALUE
+      }
 
-      return callbackMemoized(value)(...args)
-    },
-    [callbackMemoized, facetRef],
+      return callbackMemoized(...values)(...args)
+    }) as unknown as C,
```

Comment:
why do we need type assertion here? maybe we can look together into how can we avoid it?

Reply:
We have just done a pairing session with @xaviervia @prog666 and we have the fix at e1e52fe.